File: ios/SsdidWallet/Domain/MDoc/MDocPresenter.swift
import Foundation

/// Selective disclosure presenter for mdoc credentials.
///
/// Filters IssuerSignedItems to only include elements requested by the verifier,
/// preserving the issuerAuth intact so the verifier can still validate digests.
enum MDocPresenter {

    /// Filter an IssuerSigned structure to include only requested elements.
    ///
    /// - Parameters:
    ///   - issuerSigned: The full IssuerSigned structure with all data elements.
    ///   - requestedElements: Map of namespace to requested element identifiers.
    /// - Returns: A new IssuerSigned containing only the requested elements.
    static func present(
        issuerSigned: IssuerSigned,
        requestedElements: [String: [String]]
    ) -> IssuerSigned {
        var filteredNameSpaces = [String: [IssuerSignedItem]]()

        for (namespace, requestedIds) in requestedElements {
            guard let items = issuerSigned.nameSpaces[namespace] else { continue }
            let requestedSet = Set(requestedIds)
            let filtered = items.filter { requestedSet.contains($0.elementIdentifier) }
            filteredNameSpaces[namespace] = filtered
        }

        return IssuerSigned(
            nameSpaces: filteredNameSpaces,
            issuerAuth: issuerSigned.issuerAuth
        )
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/mdoc/MDocPresenter.kt
package my.ssdid.wallet.domain.mdoc

object MDocPresenter {

    fun present(
        issuerSigned: IssuerSigned,
        requestedElements: Map<String, List<String>>
    ): IssuerSigned {
        val filteredNameSpaces = mutableMapOf<String, List<IssuerSignedItem>>()

        for ((namespace, requestedIds) in requestedElements) {
            val items = issuerSigned.nameSpaces[namespace] ?: continue
            val filtered = items.filter { it.elementIdentifier in requestedIds }
            filteredNameSpaces[namespace] = filtered
        }

        return IssuerSigned(
            nameSpaces = filteredNameSpaces,
            issuerAuth = issuerSigned.issuerAuth
        )
    }
}
